### PR TITLE
[image_picker_for_web] migrates to package:web

### PR DESCRIPTION
**Pull request by balvinderz**
Updates the web implementation of `image_picker_for_web` to `package:web`.

### Issues

* Fixes https://github.com/flutter/flutter/issues/139751

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [X] I signed the [CLA].
- [X] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [X] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/image_picker/image_picker_for_web/example/integration_test/image_picker_for_web_test.dart [hunk 1]
```diff
@@ -1,27 +1,30 @@
...
-final Map<String, dynamic> options = <String, dynamic>{
-  'type': 'text/plain',
-  'lastModified': DateTime.utc(2017, 12, 13).millisecondsSinceEpoch,
-};
-final html.File textFile = html.File(<Uint8List>[bytes], 'hello.txt', options);
-final html.File secondTextFile =
-    html.File(<Uint8List>[otherBytes], 'secondFile.txt');
+final JSArray bytes =
+    const Utf8Encoder().convert(expectedStringContents).toJS as JSArray;
+final JSArray otherBytes =
```
Comment (balvinderz) [inline packages/image_picker/image_picker_for_web/example/integration_test/image_picker_for_web_test.dart:20]: same for this

File: packages/image_picker/image_picker_for_web/example/integration_test/image_picker_for_web_test.dart [hunk 2]
```diff
@@ -1,27 +1,30 @@
...
-final Uint8List bytes = const Utf8Encoder().convert(expectedStringContents);
-final Uint8List otherBytes = const Utf8Encoder().convert(otherStringContents);
-final Map<String, dynamic> options = <String, dynamic>{
-  'type': 'text/plain',
-  'lastModified': DateTime.utc(2017, 12, 13).millisecondsSinceEpoch,
-};
-final html.File textFile = html.File(<Uint8List>[bytes], 'hello.txt', options);
-final html.File secondTextFile =
-    html.File(<Uint8List>[otherBytes], 'secondFile.txt');
+final JSArray bytes =
```
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/image_picker/image_picker_for_web/example/integration_test/image_picker_for_web_test.dart:17]: Individual temporary suppressions should never be made at the file level, because it allows new, unrelated violations.

See https://github.com/flutter/packages/blob/bb97da8ec627466fdb45cdd36e2c0e35dee40e9e/packages/google_sign_in/google_sign_in_web/lib/src/flexible_size_html_element_view.dart#L76-L78 for the correct suppression for this particular issue.

[the remark this answers is not in the document]
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/example/integration_test/image_picker_for_web_test.dart:17]: i tried that too - https://github.com/flutter/packages/pull/5799/commits/ccf7796cb264b37e1a38f5e1833e0642dbd90eed
it was still failing. trying again

[the remark this answers is not in the document]
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/example/integration_test/image_picker_for_web_test.dart:17]: failed again @stuartmorgan

[the remark this answers is not in the document]
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/example/integration_test/image_picker_for_web_test.dart:17]: i had to add 4 (2 for each declaration) ignore to make this work

[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/image_picker/image_picker_for_web/example/integration_test/image_picker_for_web_test.dart:17]: `// ignore` operates on lines, not statements, so that's expected for a multi-line statement where each line has a violation.

[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/image_picker/image_picker_for_web/example/integration_test/image_picker_for_web_test.dart:17]: These declarations will need explanatory TODOs, as in the example I linked to. We should never be ignoring lint rules without explaining why.

[the remark this answers is not in the document]
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/example/integration_test/image_picker_for_web_test.dart:18]: Added todo's under your name @ditman because i am not a team member :P

[the remark this answers is not in the document]
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/example/integration_test/image_picker_for_web_test.dart]: this todo is not needed anymore.

**Comment by balvinderz** [comment]
@ditman can you please review this

File: packages/image_picker/image_picker_for_web/CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,9 @@
-## 3.0.2
+## 3.0.3
+
+* Migrates package and tests to `platform:web`.
+* Updates minimum supported SDK version to Flutter 3.16.0/Dart 3.2.0.
 
+## 3.0.2
```
Comment (tarrinneal) [inline packages/image_picker/image_picker_for_web/CHANGELOG.md:6]: there should be an empty line after this one.

**Review by tarrinneal — COMMENTED** [review]
Just a small nit from me for now, the web package is still something I haven't interacted with much so I'll let @ditman do the primary review.

**Comment by balvinderz** [comment]
I am on vacation till 17th march so wont be able to address any PR comments till then. Someone else can take over this PR if we want to land this sooner.

**Comment by ditman** [comment]
I'll take over, thanks @balvinderz as always! Enjoy your vacation!

File: packages/image_picker/image_picker_for_web/CHANGELOG.md (hunk 3 above)
Comment (ditman) [reply packages/image_picker/image_picker_for_web/CHANGELOG.md:6]: Fixed!

File: packages/image_picker/image_picker_for_web/example/integration_test/image_picker_for_web_test.dart [hunk 4]
```diff
@@ -69,30 +77,37 @@ void main() {
...
-          DateTime.fromMillisecondsSinceEpoch(textFile.lastModified!),
+          DateTime.fromMillisecondsSinceEpoch(textFile.lastModified),
         ));
-    expect(html.querySelector('flt-image-picker-inputs')?.children.isEmpty,
+    expect(
+        web.document
+                .querySelector('flt-image-picker-inputs')
+                ?.children
+                .length ==
+            0,
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/example/integration_test/image_picker_for_web_test.dart:87]: This formatting is funky, let's convince the formatter to do something better!

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 5]
```diff
@@ -226,65 +229,68 @@ class ImagePickerPlugin extends ImagePickerPlatform {
...
+  List<web.File>? _getFilesFromInput(web.HTMLInputElement input) {
     if (_hasOverrides) {
       return _overrides!.getMultipleFilesFromInput(input);
     }
-    return input.files;
+    final web.FileList? fileList = input.files;
+    final List<web.File> files = <web.File>[];
+    for (int i = 0; i < (fileList?.length ?? 0); i++) {
+      files.add(input.files!.item(i)!);
+    }
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:240]: I think this can be `.cast()`

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 6]
```diff
@@ -226,65 +229,68 @@ class ImagePickerPlugin extends ImagePickerPlatform {
...
-    input.onChange.first.then((html.Event event) {
-      final List<html.File>? files = _handleOnChangeEvent(event);
+    input.onchange = (web.Event event) {
+      final List<web.File>? files = _handleOnChangeEvent(event);
       if (!completer.isCompleted && files != null) {
-        completer.complete(files.map((html.File file) {
+        completer.complete(files.map((web.File file) {
           return XFile(
-            html.Url.createObjectUrl(file),
+            web.URL.createObjectURL(file.jsify()! as JSObject),
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:260]: file shouldn't need `jsify` anymore.

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 7]
```diff
@@ -226,65 +229,68 @@ class ImagePickerPlugin extends ImagePickerPlatform {
...
+  web.Element _ensureInitialized(String id) {
+    web.Element? target = web.document.querySelector('#$id');
     if (target == null) {
-      final html.Element targetElement =
-          html.Element.tag('flt-image-picker-inputs')..id = id;
-
-      html.querySelector('body')!.children.add(targetElement);
+      final web.Element targetElement =
+          web.document.createElement('flt-image-picker-inputs')..id = id;
+      final JSAny jsElement = targetElement.jsify()!;
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:292]: Same here, jsify shouldn't be needed.

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 8]
```diff
@@ -314,9 +325,10 @@ class ImagePickerPlugin extends ImagePickerPlatform {
   }
 
   /// Injects the file input element, and clicks on it
-  void _injectAndActivate(html.Element element) {
-    _target.children.clear();
-    _target.children.add(element);
+  void _injectAndActivate(web.HTMLElement element) {
+    _target.replaceChildren(<JSAny>[].toJS);
+    final JSAny jsElement = element.jsify()!;
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:330]: Another jsify()

**Review by ditman — COMMENTED** [review]
There's a few tweaks that shouldn't be needed with the current version of the JS interop (like a bunch of jsify calls); I'll clean this up super quickly.

File: packages/image_picker/image_picker_for_web/CHANGELOG.md [hunk 9]
```diff
@@ -1,11 +1,12 @@
-## NEXT
+## 3.0.3
 
-* Updates minimum supported SDK version to Flutter 3.13/Dart 3.1.
+* Migrates package and tests to `platform:web`.
```
Comment (kevmoo) [inline packages/image_picker/image_picker_for_web/CHANGELOG.md:3]:
```suggestion
* Migrates package and tests to `package:web`.
```

File: packages/image_picker/image_picker_for_web/CHANGELOG.md [hunk 10]
```diff
@@ -1,11 +1,12 @@
...
+* Updates minimum supported SDK version to Flutter 3.19.0/Dart 3.3.0.
 
 ## 3.0.2
 
-* Updates minimum supported SDK version to Flutter 3.10/Dart 3.0.
 * Removes input element after completion
+* Updates minimum supported SDK version to Flutter 3.10/Dart 3.0.
```
Comment (kevmoo) [inline packages/image_picker/image_picker_for_web/CHANGELOG.md:9]: Meh...maybe avoid unrelated change.

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart (hunk 5 above)
Comment (kevmoo) [reply packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:240]: And/or use collection syntax here!

Comment (ditman) [reply packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:240]: Can't be cast, because `web.FileList` is very very limited. I fixed this with collection syntax + extension method on `web.FileList`, looks fancy!

File: packages/image_picker/image_picker_for_web/CHANGELOG.md (hunk 9 above)
Comment (ditman) [reply packages/image_picker/image_picker_for_web/CHANGELOG.md:3]: 😅 Thanks, totally missed that one!

File: packages/image_picker/image_picker_for_web/CHANGELOG.md (hunk 10 above)
Comment (ditman) [reply packages/image_picker/image_picker_for_web/CHANGELOG.md:9]: Reverted (I think :P)

**Review by ditman — APPROVED** [review]
PTAL @tarrinneal, I think I addressed all my own concerns haha!

**Comment by ditman** [comment]
(I tried compiling the example app to Wasm, but it depends on `video_player`, which isn't migrated yet. All the errors happen in that package, which I'll finish migrating next)

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 11]
```diff
@@ -151,14 +153,16 @@ class ImagePickerPlugin extends ImagePickerPlatform {
     String? capture,
     bool multiple = false,
   }) {
-    final html.FileUploadInputElement input = createInputElement(
+    final web.HTMLInputElement input = createInputElement(
```
Comment (kevmoo) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:156]: can bump to 0.5.1 and use the ctor now!

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 12]
```diff
@@ -226,65 +230,63 @@ class ImagePickerPlugin extends ImagePickerPlatform {
...
-  html.Element _ensureInitialized(String id) {
-    html.Element? target = html.querySelector('#$id');
+  web.Element _ensureInitialized(String id) {
+    web.Element? target = web.document.querySelector('#$id');
     if (target == null) {
-      final html.Element targetElement =
-          html.Element.tag('flt-image-picker-inputs')..id = id;
-
-      html.querySelector('body')!.children.add(targetElement);
+      final web.Element targetElement =
```
Comment (kevmoo) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:286]: ditto

Comment (kevmoo) [reply packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:286]: er, nevermind.

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 13]
```diff
@@ -302,9 +304,14 @@ class ImagePickerPlugin extends ImagePickerPlatform {
       return _overrides!.createInputElement(accept, capture);
     }
 
-    final html.Element element = html.FileUploadInputElement()
-      ..accept = accept
-      ..multiple = multiple;
+    final web.HTMLInputElement element =
+        (web.document.createElement('input') as web.HTMLInputElement)
```
Comment (kevmoo) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:308]: ditto

File: packages/image_picker/image_picker_for_web/lib/src/pkg_web_tweaks.dart [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+// found in the LICENSE file.
+
+import 'package:web/web.dart' as web;
+
+/// Adds a `toList` method to [web.FileList] objects.
+extension WebFileListToDartList on web.FileList {
+  /// Converts a [web.FileList] into a [List] of [web.File].
+  ///
+  /// This method makes a copy.
+  List<web.File> get toList =>
```
Comment (kevmoo) [inline packages/image_picker/image_picker_for_web/lib/src/pkg_web_tweaks.dart:12]: meh...I'd opt for a function here. But maybe that's just my preference.

**Review by kevmoo — APPROVED** [review]

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart (hunk 11 above)
Comment (ditman) [reply packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:156]: Yep, updated to element constructors!

File: packages/image_picker/image_picker_for_web/lib/src/pkg_web_tweaks.dart (hunk 14 above)
Comment (ditman) [reply packages/image_picker/image_picker_for_web/lib/src/pkg_web_tweaks.dart:12]: I hope this eventually gets provided by package:web, as some form of iterable thing, so I don't need any of my code :P

File: packages/image_picker/image_picker_for_web/example/integration_test/image_picker_for_web_test.dart [hunk 15]
```diff
@@ -3,25 +3,27 @@
...
-final Map<String, dynamic> options = <String, dynamic>{
-  'type': 'text/plain',
-  'lastModified': DateTime.utc(2017, 12, 13).millisecondsSinceEpoch,
-};
-final html.File textFile = html.File(<Uint8List>[bytes], 'hello.txt', options);
-final html.File secondTextFile =
-    html.File(<Uint8List>[otherBytes], 'secondFile.txt');
+final web.FilePropertyBag options = web.FilePropertyBag(
+    lastModified: DateTime.utc(2017, 12, 13).millisecondsSinceEpoch)
+  ..type = 'text/plain';
```
[the remark this answers is not in the document]
Comment (ditman) [reply packages/image_picker/image_picker_for_web/example/integration_test/image_picker_for_web_test.dart]: Linked in the code so I don't forget to fix this once web 0.6.0 rolls.

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 16]
```diff
@@ -314,9 +320,9 @@ class ImagePickerPlugin extends ImagePickerPlatform {
   }
 
   /// Injects the file input element, and clicks on it
-  void _injectAndActivate(html.Element element) {
-    _target.children.clear();
-    _target.children.add(element);
+  void _injectAndActivate(web.HTMLElement element) {
+    _target.replaceChildren(<JSAny>[].toJS);
+    _target.append(element);
```
[the remark this answers is not in the document]
Comment (ditman) [reply packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:325]: But that breaks tests.... so maybe not :S

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 17]
```diff
@@ -151,14 +153,16 @@ class ImagePickerPlugin extends ImagePickerPlatform {
     String? capture,
     bool multiple = false,
   }) {
-    final html.FileUploadInputElement input = createInputElement(
+    final web.HTMLInputElement input = createInputElement(
       accept,
       capture,
       multiple: multiple,
-    ) as html.FileUploadInputElement;
+    ) as web.HTMLInputElement;
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:160]: No need to cast here, now that we can create the proper object in the `createInputElement` function.

File: packages/image_picker/image_picker_for_web/lib/src/image_resizer.dart [hunk 18]
```diff
@@ -23,68 +25,73 @@ class ImageResizer {
...
-      imageLoadCompleter.completeError(exception);
-    });
+  Future<web.HTMLImageElement> loadImage(String blobUrl) {
+    final Completer<web.HTMLImageElement> imageLoadCompleter =
+        Completer<web.HTMLImageElement>();
+    final web.HTMLImageElement imageElement = web.HTMLImageElement();
+    imageElement
+      // ignore: unsafe_html
+      ..src = blobUrl
+      ..onload = (web.Event event) {
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/src/image_resizer.dart:48]: use the `onLoad` stream, instead of the raw event handler (example above)

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 19]
```diff
@@ -226,65 +230,63 @@ class ImagePickerPlugin extends ImagePickerPlatform {
...
   }
 
   /// Monitors an <input type="file"> and returns the selected file(s).
-  Future<List<XFile>> _getSelectedXFiles(html.FileUploadInputElement input) {
+  Future<List<XFile>> _getSelectedXFiles(web.HTMLInputElement input) {
     final Completer<List<XFile>> completer = Completer<List<XFile>>();
     // Observe the input until we can return something
-    input.onChange.first.then((html.Event event) {
-      final List<html.File>? files = _handleOnChangeEvent(event);
+    input.onchange = (web.Event event) {
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:251]: Use the streams, not the listeners directly.

Comment (ditman) [reply packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:253]: Created an issue for this, the `onCancel` stream doesn't exist and seems that cannot be easily created: https://github.com/dart-lang/web/issues/199

**Review by ditman — APPROVED** [review]
Adding 2nd approval.
